Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm

```diff
@@ -10965,23 +10958,71 @@ - (void)imageAnalysisGestureDidBegin:(WKImageAnalysisGestureRecognizer *)gesture
     } forRequest:request];
 }
 
+static void collectAllActionIdentifiers(UIMenu *menu, NSMutableSet<UIActionIdentifier> *actions)
+{
+    for (UIMenuElement *child in menu.children) {
+        if (auto identifier = dynamic_objc_cast<UIAction>(child).identifier)
+            [actions addObject:identifier];
+        else if (auto subMenu = dynamic_objc_cast<UIMenu>(child))
+            collectAllActionIdentifiers(subMenu, actions);
+    }
+}
+
 - (void)_completeImageAnalysisRequestForContextMenu:(CGImageRef)image requestIdentifier:(WebKit::ImageAnalysisRequestIdentifier)requestIdentifier hasTextResults:(BOOL)hasTextResults
 {
-#if USE(QUICK_LOOK)
-    _hasSelectableTextInImage = hasTextResults;
-#endif
+    [self _invokeAllActionsToPerformAfterPendingImageAnalysis:WebKit::ProceedWithTextSelectionInImage::No];
+
+    auto data = Box<WebKit::ImageAnalysisContextMenuActionData>::create();
+    data->hasSelectableText = hasTextResults;
 
     auto weakSelf = WeakObjCPtr<WKContentView>(self);
-    auto aggregator = CallbackAggregator::create([weakSelf, requestIdentifier]() mutable {
-        if (auto strongSelf = weakSelf.get(); [strongSelf validateImageAnalysisRequestIdentifier:requestIdentifier])
-            [strongSelf _invokeAllActionsToPerformAfterPendingImageAnalysis:WebKit::ProceedWithTextSelectionInImage::No];
+    auto aggregator = CallbackAggregator::create([weakSelf, data]() mutable {
+        auto strongSelf = weakSelf.get();
+        if (!strongSelf)
+            return;
+
+        strongSelf->_imageAnalysisContextMenuActionData = WTFMove(*data);
+        [[strongSelf contextMenuInteraction] updateVisibleMenuWithBlock:makeBlockPtr([strongSelf](UIMenu *menu) -> UIMenu * {
+            auto adjustedChildren = adoptNS(menu.children.mutableCopy);
+
+            auto *existingActionIdentifiers = [NSMutableSet<NSString *> set];
+            collectAllActionIdentifiers(menu, existingActionIdentifiers);
+
+            auto *elementInfo = [_WKActivatedElementInfo activatedElementInfoWithInteractionInformationAtPosition:strongSelf->_positionInformation userInfo:nil];
+            auto addActionIfNeeded = [&](_WKElementActionType action) {
+                auto actionIdentifier = elementActionTypeToUIActionIdentifier(action);
+                if (!actionIdentifier || [existingActionIdentifiers containsObject:actionIdentifier])
```

Comment:
Instead of looking through the existing actions to see if we already have one that matches, what about if we have some extra `BOOL` inside the `ImageAnalysisContextMenuActionData` that indicate whether or not we should add the relevant item to the contextmenu?

e.g. in addition to `hasSelectableTextForImageContextMenu `, we could have a `shouldAddImageExtractionContextMenuItem` (and maybe a `didAddImageExtractionContextMenuItem`) that is called anytime we're about to add a `_WKElementActionTypeImageExtraction` item

Reply:
After discussing this over Slack, and agreed on a solution where we:
- Add something to tell if the context menu has been overridden via delegate
- if so, then don’t add these image analysis actions at all
- if not, then add them without trying to mitigate duplicate IDs